### PR TITLE
OSD-10987 - Adding tests for RPO

### PR DESCRIPTION
**Pull request by ravitri**
### What type of PR is this?
_(test)_

### What this PR does / why we need it?
PR adds more test coverage

### Which Jira/Github issue(s) this PR fixes?
_Fixes #_ [OSD-10987](https://issues.redhat.com/browse/OSD-10987)

### Pre-checks (if applicable):
- [ ] Tested latest changes against a cluster
- [ ] Included documentation changes with PR

### CODE REVIEW

**Comment by ravitri** [comment]
/hold

The PR can be reviewed for now. Some more things to be added in coming week. Meanwhile, adding hold.

File: pkg/const/test/test.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+			},
+		},
+	}
+
+	TestAllowedList = "default"
+
+	TestDeniedList = ""
```
Comment (mrbarge) [inline pkg/const/test/test.go:129]: Minor nits - making these names more descriptive helps give better context for the tests in which they're used (particularly the deny one).

ie something like
```suggestion
	TestDefaultAllowedList = "default"

	TestEmptyDeniedList = ""
```

File: pkg/controller/utils/controllerutil_test.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	})
+
+	Context("Running GenerateSafeList", func() {
+
+		It("Should return safe list if the deny list is blank", func() {
+			safeList := GenerateSafeList(testconst.TestAllowedList, testconst.TestDeniedList, testconst.TestNamespaceList)
+			Expect(safeList).To(ContainElement(ContainSubstring("default.whatever")))
+		})
+
+		It("Should not return any list if the deny list is same as safe list", func() {
```
Comment (mrbarge) [inline pkg/controller/utils/controllerutil_test.go:77]:
```suggestion
		It("Should not return any list if the deny list is same as allow list", func() {
```

File: pkg/controller/subjectpermission/subjectpermission_controller_test.go [hunk 3]
```diff
@@ -1,420 +1,613 @@
-package subjectpermission
+package subjectpermission_test
```
[the remark this answers is not in the document]
Comment (ravitri) [reply pkg/controller/subjectpermission/subjectpermission_controller_test.go:1]: Agreed. Have added the required tests.

File: pkg/controller/namespace/namespace_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+				},
+			}
+			gomock.InOrder(
+				mockClient.EXPECT().Get(gomock.Any(), testconst.TestNamespaceName, gomock.Any()).Times(1).SetArg(2, *testNamespace),
+				mockClient.EXPECT().List(gomock.Any(), gomock.Any()).Times(1).SetArg(1, *testconst.TestNamespaceList),
+				mockClient.EXPECT().List(gomock.Any(), gomock.Any()).Times(1).SetArg(1, testSubjectPermissionList),
+				mockClient.EXPECT().List(gomock.Any(), gomock.Any(), []client.ListOption{
+					client.InNamespace(testconst.TestNamespaceName.Name),
+				}).Times(1).SetArg(1, *testconst.TestRoleBindingList),
+				mockClient.EXPECT().Create(gomock.Any(), gomock.Any()).SetArg(1, *testconst.TestRoleBinding),
```
[the remark this answers is not in the document]
Comment (ravitri) [reply pkg/controller/namespace/namespace_controller_test.go]: Thanks Matt.. Yeah, agreed.. I have improved the tests with more validations now with `DoAndReturn` for `Create` and `Update` tests.

File: pkg/controller/subjectpermission/subjectpermission_controller_test.go [hunk 5]
```diff
@@ -1,420 +1,613 @@
...
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("Updates status condition if any ClusterRoleName does not exist as a ClusterRole", func() {
+			gomock.InOrder(
+				mockClient.EXPECT().Get(gomock.Any(), testconst.TestNamespaceName, gomock.Any()).Times(1).SetArg(2, testconst.TestSubjectPermission),
+				mockClient.EXPECT().List(gomock.Any(), gomock.Any()).Times(1).SetArg(1, testconst.TestClusterRoleList),
+				mockClient.EXPECT().List(gomock.Any(), gomock.Any()).Times(1).SetArg(1, testconst.TestClusterRoleBindingList),
+				mockClient.EXPECT().Status().Return(mockStatusWriter),
+				mockStatusWriter.EXPECT().Update(gomock.Any(), gomock.Any()).Times(1).SetArg(1, testconst.TestSubjectPermission),
```
[the remark this answers is not in the document]
Comment (ravitri) [reply pkg/controller/subjectpermission/subjectpermission_controller_test.go]: Done

File: pkg/controller/subjectpermission/subjectpermission_controller_test.go [hunk 6]
```diff
@@ -1,420 +1,613 @@
...
+		mockStatusWriter = clientmocks.NewMockStatusWriter(mockCtrl)
+		subjectPermissionReconciler = subjectpermission.ReconcileSubjectPermission{
+			Client: mockClient,
+			Scheme: testconst.Scheme,
+		}
+	})
+
+	Context("Reconciling SubjectPermission", func() {
+
+		It("Should clear Prometheus metrics for SubjectPermission if the deletion timestamp is not nil", func() {
```
[the remark this answers is not in the document]
Comment (ravitri) [reply pkg/controller/subjectpermission/subjectpermission_controller_test.go]: Agreed.. Considering that RPO doesn't have alerts for now, I will postpone adding the tests for the metrics for a later time.. For now, I will prefer to skip this..

File: pkg/controller/subjectpermission/subjectpermission_controller_test.go [hunk 7]
```diff
@@ -1,420 +1,613 @@
...
+				mockClient.EXPECT().List(gomock.Any(), gomock.Any()).Times(1).SetArg(1, testconst.TestClusterRoleBindingList),
+				mockClient.EXPECT().Create(gomock.Any(), gomock.Any()).Times(2).SetArg(1, testconst.TestClusterRoleBinding),
+				mockClient.EXPECT().Status().Return(mockStatusWriter),
+				mockStatusWriter.EXPECT().Update(gomock.Any(), gomock.Any()).SetArg(1, testconst.TestSubjectPermission),
+			)
+			_, err := subjectPermissionReconciler.Reconcile(testconst.Context, reconcile.Request{NamespacedName: testconst.TestNamespaceName})
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("Should successfully reconcile if there are no ClusterPermissions and no safe namespaces for rolebinding creation", func() {
```
[the remark this answers is not in the document]
Comment (ravitri) [reply pkg/controller/subjectpermission/subjectpermission_controller_test.go]: Thanks Matt.. I have split the description of the tests between `When` and `It` and it seems bit better after that I think.. Let me know if you have more suggestions for it..

**Comment by mrbarge** [comment]
Thanks for implementing all those suggestions. I've re-read through it and it's great! 
/approve

**Pull request merged**